Core/Movement: Corrected Animation Tier handling

Faq opened the pull request:
git cherry-pick 5cb747620a7cb897135650c519ca2daa50523f40

Conflicts:
src/server/game/Entities/Creature/Creature.cpp
src/server/game/Entities/Unit/Unit.cpp
src/server/game/Entities/Unit/Unit.h
src/server/game/Movement/MotionMaster.cpp
src/server/game/Movement/Spline/MoveSplineFlag.h
src/server/game/Movement/Spline/MoveSplineInit.h
src/server/scripts/EasternKingdoms/BastionOfTwilight/boss_theralion_and_valiona.cpp
src/server/scripts/EasternKingdoms/BlackrockMountain/BlackwingDescent/boss_atramedes.cpp
src/server/scripts/EasternKingdoms/BlackrockMountain/BlackwingDescent/boss_nefarians_end.cpp
src/server/scripts/EasternKingdoms/BlackrockMountain/BlackwingDescent/instance_blackwing_descent.cpp
src/server/scripts/EasternKingdoms/Gilneas/gilneas_chapter_1.cpp
src/server/scripts/EasternKingdoms/GrimBatol/boss_drahga_shadowburner.cpp
src/server/scripts/EasternKingdoms/GrimBatol/grim_batol.cpp
src/server/scripts/Kalimdor/CavernsOfTime/DragonSoul/boss_madness_of_deathwing.cpp
src/server/scripts/Kalimdor/VortexPinnacle/boss_altairus.cpp
src/server/scripts/Kalimdor/VortexPinnacle/vortex_pinnacle.cpp
src/server/scripts/Maelstrom/Stonecore/boss_slabhide.cpp
src/server/scripts/Northrend/Nexus/EyeOfEternity/boss_malygos.cpp
src/server/scripts/Northrend/Ulduar/Ulduar/boss_mimiron.cpp

<!--- (**********************************)
      (** Fill in the following fields **)
      (**********************************) --->

**Changes proposed:**

-  described in #22421 and https://github.com/The-Cataclysm-Preservation-Project/TrinityCore/pull/113

**Target branch(es):** 3.3.5/master

- [x] 3.3.5

**Issues addressed:** Closes #22421 and #25281


**Tests performed:** (Does it build, tested in-game, etc.)
Compiles and checked all mentioned in todo
**Known issues and TODO list:** (add/remove lines as needed)
Check scripts in files where code changed:
- [x] boss_nightbane
- [x] boss_onyxia
- [x] boss_jedoga_shadowseeker
- [x] boss_saviana_ragefire
- [x] boss_blood_queen_lana_thel
- [x] Infiltrator Minchar landing part, bug existing before PR #25281
- [x] boss_sindragosa
- [x] boss_the_lich_king
- [x] boss_malygos
- [x] boss_mimiron
- [x] boss_razorscale
- [x] npc_enslaved_proto_drake (utgarde_keep)
- [x] boss_skadi
- [x] npc_enslaved_netherwing_drake (zone_shadowmoon_valley)


<!--- Notes
- Enable the setting "[√] Allow edits from maintainers." when creating your pull request.
- If this PR only contains SQL files, open a new issue instead and post or link the SQL in the issue.
- When adding new SQL files, name them 9999_99_99_99_db_name.sql to reduce the chance of possible merge conflicts.
--->

Warpten committed:
Core/Movement: Fixed Animation Tier handling

# Conflicts:
#	src/server/game/Entities/Creature/Creature.cpp
#	src/server/game/Entities/Unit/Unit.cpp
#	src/server/game/Entities/Unit/Unit.h
#	src/server/game/Movement/MotionMaster.cpp
#	src/server/game/Movement/Spline/MoveSplineFlag.h
#	src/server/game/Movement/Spline/MoveSplineInit.h
#	src/server/scripts/EasternKingdoms/BastionOfTwilight/boss_theralion_and_valiona.cpp
#	src/server/scripts/EasternKingdoms/BlackrockMountain/BlackwingDescent/boss_atramedes.cpp
#	src/server/scripts/EasternKingdoms/BlackrockMountain/BlackwingDescent/boss_nefarians_end.cpp
#	src/server/scripts/EasternKingdoms/BlackrockMountain/BlackwingDescent/instance_blackwing_descent.cpp
#	src/server/scripts/EasternKingdoms/Gilneas/gilneas_chapter_1.cpp
#	src/server/scripts/EasternKingdoms/GrimBatol/boss_drahga_shadowburner.cpp
#	src/server/scripts/EasternKingdoms/GrimBatol/grim_batol.cpp
#	src/server/scripts/Kalimdor/CavernsOfTime/DragonSoul/boss_madness_of_deathwing.cpp
#	src/server/scripts/Kalimdor/VortexPinnacle/boss_altairus.cpp
#	src/server/scripts/Kalimdor/VortexPinnacle/vortex_pinnacle.cpp
#	src/server/scripts/Maelstrom/Stonecore/boss_slabhide.cpp
#	src/server/scripts/Northrend/Nexus/EyeOfEternity/boss_malygos.cpp
#	src/server/scripts/Northrend/Ulduar/Ulduar/boss_mimiron.cpp

Faq commented:
@Warpten you have plans with this?

Aokromes commented:
TrinityCore/src/server/game/Entities/Creature/Creature.cpp:2888:80: fatal error: 
      unused parameter 'updateAnimationTier' [-Wunused-parameter]
  ...disable, bool packetOnly/*=false*/, bool updateAnimationTier/*= true*/)
                                              ^

Loopeless commented:
@Faq , would this be valid for master (8.3.x) branch also or only 3.3.5.

Aokromes commented:
> @Faq , would this be valid for master (8.3.x) branch also or only 3.3.5.

it's pick from master branch.

Shauren commented:
No, this comes from 4.3.4

Loopeless commented:
The reason why I asked is because I saw @Shauren conversation with xvwyh and he said this:

> Yeah, I was looking at 8.1.5 client - there is no updatefield handler for animTier and similar code is in handler for SMSG_SET_ANIM_TIER (reading value from packet instead of updatefield)

So it made me wonder if this one is valid at all for master branch... and yeah I know it comes from 4.3.4

Shauren commented on src/server/game/Movement/Spline/MoveSplineFlag.h:
```diff
@@ -58,8 +58,8 @@ namespace Movement
-            // animation ids stored here, see AnimType enum, used with Animation flag
-            Mask_Animations     = 0xFF,
+            // animation ids stored here, see AimationTier enum, used with Animation flag
+            Mask_Animations     = 0x7,
```
This value is supposed to cover all bits specified below for animId, 8 bits means 0xFF

Faq replied:
Thnx, changed.

Ovahlord committed:
Core/Movement: Handle anim tiers inside of the Creature class.

* no longer set anim tiers when the target is rooted

Carbenium committed:
DB/Creature: Disable automatic movement flag updates for Grauf

We need manual control of fly/ground states in the script
since Graufs' spawn position is above the ground floor (sniffed!)
and thus the automatic movement flag system would enable
fly animations already at spawn.

Warpten commented on src/server/game/Entities/Creature/Creature.cpp:
```diff
@@ -1986,8 +1987,9 @@ void Creature::setDeathState(DeathState s)
-        SetHover(false);
+        bool needsFalling = (IsFlying() || IsHovering()) && !IsUnderWater();
+        SetHover(false, false);
+        SetDisableGravity(false, false);
```
Why

Carbenium replied:
This one is present in your 434 commit as well. https://github.com/TrinityCore/TrinityCore/commit/5cb747620a7cb897135650c519ca2daa50523f40#diff-f3f953a63e2a008d22de69cb68f735c4R1827
:shrug:

Warpten replied:
Well that's an oops? @Ovahlord unless this was the corpse falling issue? I forgot already

Ovahlord replied:
pretty sure it was because of corpse falling, ye

Warpten commented on src/server/scripts/Northrend/AzjolNerub/Ahnkahet/boss_jedoga_shadowseeker.cpp:
```diff
@@ -246,7 +246,6 @@ struct boss_jedoga_shadowseeker : public BossAI
             if (++_initiatesKilled == TWILIGHT_INITIATES_SIZE)
             {
                 DoCastSelf(SPELL_HOVER_FALL_1);
-                me->RemoveByteFlag(UNIT_FIELD_BYTES_1, UNIT_BYTES_1_OFFSET_ANIM_TIER, UNIT_BYTE1_FLAG_ALWAYS_STAND | UNIT_BYTE1_FLAG_HOVER);
```
```suggestion
                SetAnimationTier(AnimationTier::Ground);
```

Faq replied:
tested, ok

Warpten commented on src/server/scripts/Northrend/IcecrownCitadel/boss_blood_queen_lana_thel.cpp:
```diff
@@ -215,7 +215,6 @@ class boss_blood_queen_lana_thel : public CreatureScript
                     if (Creature* minchar = me->FindNearestCreature(NPC_INFILTRATOR_MINCHAR_BQ, 200.0f))
                     {
                         minchar->SetUInt32Value(UNIT_NPC_EMOTESTATE, 0);
-                        minchar->RemoveByteFlag(UNIT_FIELD_BYTES_1, UNIT_BYTES_1_OFFSET_ANIM_TIER, UNIT_BYTE1_FLAG_ALWAYS_STAND);
```
Replace with  `minchar->SetAnimationTier(AnimationTier::Ground);`

Faq replied:
tested, ok

Warpten replied:
If I recall correctly his spawn coordinates are in the air? If so, could be caused by https://github.com/TrinityCore/TrinityCore/pull/24875/files#diff-f3f953a63e2a008d22de69cb68f735c4R1992

Warpten replied:
Why did you mark this as resolved?
> he is not on correct position while she is alive (he is on ground of first floor not over her head).

Faq replied:
added in todo

Warpten commented:
Test Lana'thel's Delirious Slash and test minchar quest

Faq commented:
This PR is mess now with all the forced pushes from @Carbenium, cant even accept suggested changes.

Carbenium commented:
@Faq That has nothing to do with it. It's a Github limitation that you can't accept suggestions on deleted lines.

Faq committed:
Feedback corrections

Faq commented:
Will recheck all where was removed some code part, todo list updated.

Ovahlord commented:
7c5c4bd9fa867d9e2fdf50712f596222e7666ca9  one should know that scripted creatures with controlled flight movement is not suposed to update its movement flags on its on neither should creatures should start with flight = 1 since that's a hack. Give her ignore pathfinding instead if the shortcut is a problem.

Faq committed:
DB/Creature: Correcting movement flags for Blood-Queen Lana'thel

from 7c5c4bd9fa867d9e2fdf50712f596222e7666ca9

Faq committed:
DB/Creature: Correcting movement flags for Malygos

Faq committed:
DB/Creature: More movement flag changes
Not allow to be half way in ground.

Carbenium commented on sql/updates/world/3.3.5/2020_08_xx_xx_world.sql:
```diff
@@ -0,0 +1,15 @@
+UPDATE `creature_template` SET `flags_extra` = `flags_extra`|512 WHERE `entry`=26893;
+-- Disable movement flag updates for Blood-Queen Lana'thel
+UPDATE `creature_template_movement` SET `Ground`= 1, `Flight`= 0 WHERE `CreatureId` IN (37955, 38434, 38435, 38436);
+UPDATE `creature_template` SET `flags_extra`= `flags_extra` | 513 WHERE `entry` IN (37955, 38434, 38435, 38436);
```
Should this really be `513` and not `512`?

Faq replied:
it had 1 before so with that flag =513, not all have 1=INSTANCE_BIND

Carbenium replied:
You are OR-ing the value. `flags_extra | 512` means take the existing value and "add" 512. The end-result will be 513 if that flag was already set. -> Change those 513 values to 512 (or even better 0x200).

Faq committed:
Correcting

Faq committed:
Scripts/mimiron: Put head (Aerial Command Unit) on shoulders

Without this, head is too low (50% below shoulders).

Faq commented:
All checks done, ready from my side.

Faq committed:
Scripts/Creature: Infiltrator Minchar
He is now in air at start position

jackpoz commented:
was this tested ingame too ?

Faq commented:
Wholo todo was done ingame. More questions?

Faq committed:
Merge branch '3.3.5' into Cherry3

# Conflicts:
#	src/server/scripts/Northrend/IcecrownCitadel/boss_sindragosa.cpp
#	src/server/scripts/Northrend/IcecrownCitadel/boss_the_lich_king.cpp

jackpoz commented:
are the merge conflicts from last merge ok ? I see a lot of deleted code

Faq commented:
yeah, those are ok.

Faq commented:
Why this still not merged?

jackpoz commented:
@Ovahlord can you take a look at this ?

Ovahlord committed:
Merge branch '3.3.5' into Cherry3

Treeston commented on src/server/game/Entities/Player/Player.cpp:
```diff
@@ -26618,7 +26614,7 @@ bool Player::SetCanFly(bool apply, bool packetOnly /*= false*/)
-bool Player::SetHover(bool apply, bool packetOnly /*= false*/)
+bool Player::SetHover(bool apply, bool packetOnly /*= false*/, bool /*updateAnimationTier = true*/)
 {
     if (!packetOnly && !Unit::SetHover(apply))
```
Should this forward `updateAnimationTier` to `Unit::SetHover`?

Warpten replied:
Unit doesn't use the parameter but sure I guess.

Treeston commented on src/server/game/Entities/Player/Player.cpp:
```diff
@@ -26579,7 +26575,7 @@ bool Player::IsInWhisperWhiteList(ObjectGuid guid)
-bool Player::SetDisableGravity(bool disable, bool packetOnly /*= false*/)
+bool Player::SetDisableGravity(bool disable, bool packetOnly /*= false*/, bool /*updateAnimationTier = true*/)
 {
     if (!packetOnly && !Unit::SetDisableGravity(disable))
```
same question

Treeston commented on src/server/game/Entities/Unit/Unit.cpp:
```diff
@@ -513,8 +513,13 @@ void Unit::UpdateSplineMovement(uint32 t_diff)
         DisableSpline();
 
+        if (movespline->HasAnimation())
+            SetAnimationTier(movespline->GetAnimation());
```
`GetAnimationTier` perhaps?

Ovahlord replied:
I think that's to follow the packet name and movespline setter helper
`movementSpline.Animation->AnimTier = splineFlags.getAnimationId();`

Treeston commented on src/server/scripts/Northrend/Ulduar/Ulduar/boss_mimiron.cpp:
```diff
@@ -991,7 +990,7 @@ class boss_vx_001 : public CreatureScript
                         me->RemoveAurasDueToSpell(SPELL_FREEZE_ANIM);
                         me->SetUInt32Value(UNIT_NPC_EMOTESTATE, EMOTE_ONESHOT_NONE); // Remove emotestate.
-                        //me->SetUInt32Value(UNIT_FIELD_BYTES_1, UNIT_BYTES_1_OFFSET_ANIM_TIER, UNIT_BYTE1_FLAG_ALWAYS_STAND | UNIT_BYTE1_FLAG_HOVER); Blizzard handles hover animation like this it seems.
+                        //me->SetHover(true); // Blizzard handles hover animation like this it seems.
```
should this still be commented?

Keader replied:
we have hover implemented now, but mimiron needs a rewrite xd

Treeston reviewed (commented):
were all the script changes tested?

Treeston commented on src/server/game/Entities/Creature/Creature.cpp:
```diff
@@ -2884,13 +2887,23 @@ bool Creature::SetWalk(bool enable)
 {
     //! It's possible only a packet is sent but moveflags are not updated
     //! Need more research on this
     if (!packetOnly && !Unit::SetDisableGravity(disable))
```
Here, too.

(And I know it doesn't use it, but it's still good practice in case it ever gets changed.)

Ovahlord committed:
nitpicking part one

Ovahlord committed:
nitpicking part 2

Ovahlord committed:
fixed build

Treeston reviewed (approved):
get six green checks then you're good

Jildor commented:
What testing has been done on this PR before be merged?

Now Onyxia does not walk on the ground in third phase, she stays flying on the ground

jackpoz commented:
@Ovahlord is there anything to do in https://github.com/TrinityCore/TrinityCore/pull/24875/files#diff-1accb5382290646a614a61b9e06892e0 to fix Onyxia ?

Faq commented:
Will check it today
